File: domain/src/main/kotlin/com/kyovo/domain/port/secondary/BookingRepository.kt
package com.kyovo.domain.port.secondary

import com.kyovo.domain.model.booking.*
import com.kyovo.domain.model.room.RoomId
import com.kyovo.domain.model.user.UserId

interface BookingRepository
{
    fun findAll(): List<Booking>

    fun findById(id: BookingId): Booking?

    fun findByUserId(userId: UserId): List<Booking>

    fun save(booking: Booking): Booking

    fun update(booking: Booking): Booking

    fun existsOverlappingBooking(roomId: RoomId, startDate: BookingStartDate, endDate: BookingEndDate): Boolean

    fun saveStatusHistory(
        bookingId: BookingId,
        status: BookingStatus,
        changedAt: BookingStatusHistoryDate,
        changedBy: UserId?,
        reason: BookingStatusReason?
    )

    fun findStatusHistory(bookingId: BookingId): List<BookingStatusHistory>
}

File: frontend/src/domain/port/secondary/BookingRepository.ts
import type { Booking, NewBooking } from '@/domain/model/Booking'

export interface BookingRepository {
  findAll(): Promise<Booking[]>
  findByUser(userId: string): Promise<Booking[]>
  create(booking: NewBooking): Promise<Booking>
  cancel(id: string, userId: string, reason?: string): Promise<void>
}
